src/Language/Java/Inline.hs

```diff
@@ -399,8 +401,39 @@ blockQQ input = case Java.parser Java.block input of
       -- returning boxed values. Once this limitation of the compiler gets
       -- lifted, we'll support returning unboxed values, just like `call` does.
       castReturnType funcall = [| unsafeUncoerce . coerce <$> $funcall |]
-      unfoldHeadTySyn ty@(TH.ConT nT) = do
-        TH.reify nT >>= \case
-          TH.TyConI (TH.TySynD _ _ ty') -> unfoldHeadTySyn ty'
-          _ -> return ty
-      unfoldHeadTySyn ty = return ty
+
+-- Recursively unfold type synonyms, if any.
+unfoldTypeTySyn :: TH.Type -> Q TH.Type
+unfoldTypeTySyn = \case
+    -- The workhorse
+    ty@(TH.ConT name) ->
+      TH.reify name >>= \case
+        TH.TyConI (TH.TySynD _ _ ty') -> unfoldTypeTySyn ty'
```

Comment:
Good point. I think for the purposes of this patch, supporting parameterized type synonyms is not required, but that seems like a desirable feature to eventually support.